io.openems.edge.goodwe.et/src/io/openems/edge/goodwe/et/ess/GoodWeEtBatteryInverterImpl.java

```diff
@@ -507,19 +505,49 @@ private void updatechannels() {
 		/*
 		 * Update Allowed charge and Allowed discharge
 		 */
+
+//		System.out.println(
+//				"Allowed Charge power: " + allowedChargePower + " Allowed Discharge Power: " + allowedDischargePower);
+
 		Integer soc = this.getSoc().get();
-		Integer maxApparentPower = this.getMaxApparentPower().get();
+		Integer maxBatteryPower = this.config.maxBatteryPower();
 
-		if (soc == null || soc >= 99) {
-			this._setAllowedChargePower(0);
-		} else {
-			this._setAllowedChargePower(TypeUtils.multiply(maxApparentPower, -1));
+		Integer allowedCharge = null;
+		Integer allowedDischarge = null;
+
+		if (productionPower == null) {
+			productionPower = 0;
 		}
-		if (soc == null || soc <= 0) {
-			this._setAllowedDischargePower(0);
-		} else {
-			this._setAllowedDischargePower(maxApparentPower);
+
+		if (soc == null) {
+
+			allowedCharge = 0;
+			allowedDischarge = 0;
+
+		} else if (soc == 100) {
+
+			allowedDischarge = maxBatteryPower + productionPower;
+			allowedCharge = 0;
+
+		} else if (soc > 0) {
+
+			allowedDischarge = maxBatteryPower + productionPower;
+			allowedCharge = maxBatteryPower;
+
+		} else if (soc == 0) {
+
+			allowedDischarge = productionPower;
+			allowedCharge = maxBatteryPower;
+
+		}
+
+		// to avoid charging when production is greater than maximum battery power.
+		if (allowedCharge < 0) {
```

Comment:
@venu-sagar: this is another (theoretical) NullPointerException.
If you would initialize allowedCharge instead of
`Integer allowedCharge = null;`
with
`final Integer allowedCharge`
Eclipse IDE would give you a warning.